src/DictDownloadManager.cs

```diff
@@ -0,0 +1,110 @@
+﻿using Flow.Launcher.Plugin;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Dictionary
+{
+    public class DictDownloadManager
+    {
+        private string dictPath;
+        private bool downloading = false;
+        private int downloadPercentage = 0;
+        private PluginInitContext context;
+
+        public DictDownloadManager(string dictPath, PluginInitContext context)
+        {
+            this.dictPath = dictPath;
+            this.context = context;
+        }
+
+        private async Task<bool> CheckForGoogleConnection()
+        {
+            try
+            {
+                var request = WebRequest.Create("https://google.com/generate_204");
+                request.Timeout = 2000;
+                await request.GetResponseAsync();
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        private async Task<string> GetDownloadUrl()
+        {
+            bool shouldUseMirror = !await CheckForGoogleConnection();
+            if (shouldUseMirror)
+                return "https://nullptr_t.coding.net/p/ECDICT-sqlite/d/ECDICT-sqlite/git/raw/master/ecdict-ultimate-sqlite.zip";
+            else
+                return "https://github.com/skywind3000/ECDICT-ultimate/releases/download/1.0.0/ecdict-ultimate-sqlite.zip";
+        }
+
+        public async void PerformDownload()
+        {
+            downloading = true;
+
+            var url = await GetDownloadUrl();
+            var path = Path.GetDirectoryName(dictPath);
+            if (!Directory.Exists(path)) Directory.CreateDirectory(path);
+
+            var client = new WebClient();
+
+            client.DownloadProgressChanged += delegate (object sender, DownloadProgressChangedEventArgs e)
+            {
+                downloadPercentage = e.ProgressPercentage;
+            };
+
+            var tmpDictFileLoc = dictPath + ".download";
+            await client.DownloadFileTaskAsync(new Uri(url), tmpDictFileLoc).ConfigureAwait(false);
+
+            await Task.Run(() => ZipFile.ExtractToDirectory(tmpDictFileLoc, Path.GetDirectoryName(dictPath)));
+
+            File.Delete(tmpDictFileLoc);
+
+            downloading = false;
+        }
+
+        public async Task<List<Result>> HandleQueryAsync(Query query)
+        {
+            if (downloading)
+            {
+                var progress = "";
+                if (downloadPercentage != 0) progress = $"{progress} %";
```

Comment:
Shouldn't this should be downloadPercentage?

Reply:
Ah you're right. Thanks!